src/iterative_ensemble_smoother/_iterative_ensemble_smoother.py

```diff
@@ -85,70 +127,87 @@ def fit(
 
+        if ensemble_mask is None:
+            ensemble_mask = np.ones(ensemble_size, dtype=bool)
+        self.ensemble_mask = ensemble_mask
+
         # If it's the first time the method is called, create coeff matrix
         if not hasattr(self, "coefficient_matrix"):
             self.coefficient_matrix = np.zeros(shape=(ensemble_size, ensemble_size))
 
-        # Seed here so we draw the same samples in every iteration (call to update())
-        rng = np.random.default_rng(self.seed)
-
-        # A covariance matrix was passed
-        # Columns of E should be sampled from N(0, Cdd) and centered, Evensen 2019
-        if observation_errors.ndim == 2:
-            E = rng.multivariate_normal(
-                mean=np.zeros_like(observation_values),
-                cov=observation_errors,
-                size=ensemble_size,
-                method="cholesky",  # An order of magnitude faster than 'svd'
-            ).T
-        # A vector of standard deviations was passed
-        else:
-            E = rng.normal(
-                loc=0, scale=observation_errors, size=(ensemble_size, num_obs)
-            ).T
+        # ---------------------------------------------------------------------
+        # ----------- Computations corresponding to algorithm 1 ---------------
+        # ---------------------------------------------------------------------
 
+        # Draw samples from N(0, C_dd)
+        E = self._get_E(
+            observation_errors=observation_errors,
+            observation_values=observation_values,
+            ensemble_size=ensemble_size,
+        )
         assert E.shape == (num_obs, ensemble_size)
 
-        E -= E.mean(axis=1, keepdims=True)
-
-        R, observation_errors_std = _create_errors(observation_errors, inversion)
+        # Transform covariance matrix to correlation matrix
+        R, observation_errors_std = covariance_to_correlation(observation_errors)
 
         # Store D as defined by Equation (14) in Evensen (2019)
         self.D_ = E + observation_values.reshape(num_obs, 1)
         D = self.D_ - response_ensemble
 
+        # Note on scaling
+        # -------------------
+        # In line 8 in Algorithm 1, we have to compute (S S^T + E E^T)^-1 H, where
+        # H := (SW + D - g(X)). This is equivalent to solving the following
+        # equation for an unknown matrix M:
+        #     (S S^T + E E^T) M = (SW + D - g(X))
+        # Afterwards we compute S.T M. If we scale the rows (observed variables)
+        # of these equations using the standard deviations, we can obtain better
+        # conditioning numbers on the equation. This corresponds to left-multiplying
+        # with a diagonal matrix L := sqrt(diag(C_dd)). In an experiment with
+        # random covariance matrices, this improved the condition number ~90%
+        # of the time (results may depend on how random covariance matrices are
+        # generated --- I covariances C by first E ~ stdnorm(), then C = E.T @ E).
```

Comment:
Did you perhaps mean to write "I generated covariances C ..."?

Reply:
yes, will fix!